RFC: Adds speculative API specification

despeset opened the pull request:
Wildly speculative API for a micro-framework -- not to be merged ( at least, until the API is fully implemented :bowtie: )

Discuss.

despeset committed:
Adds speculative API specification

despeset committed:
Adds speculative API specification

zachallaun commented:
I'M FOR IT.

despeset commented on example-app.jl:
```diff
@@ -0,0 +1,68 @@
+    req.resources[:id] == 99 ? string("Go to hell ", req.resources[:name]) : string("Hello ", req.resources[:name])
+end
+
+route(app, GET, "/") do req, res
```
Does this `GET` stynax qualify as Type abuse?  Would it be better if it were a symbol?

StefanKarpinski replied:
I don't know what GET is, so hard to say. It could just be an integer. Or it could be a type object that's being dispatched on.

zachallaun replied:
My thought was that this was an integer.
On Mar 19, 2013 11:30 PM, "Stefan Karpinski" notifications@github.com
wrote:

> In example-app.jl:
> 
> > +
> > +# Route with HTTP verbs
> > +get(app, "/") do req, res
> > -    "hello"
> >   +end
> >   +
> >   +post(app, "/") do req, res
> > -    # ...
> >   +end
> >   +
> >   +# Route with URL vars
> >   +route(app, "/hello/id::Int/name::%[a-z]{3}[0-9]{2}") do req, res
> > -    req.resources[:id] == 99 ? string("Go to hell ", req.resources[:name]) : string("Hello ", req.resources[:name])
> >   +end
> >   +
> >   +route(app, GET, "/") do req, res
> 
> I don't know what GET is, so hard to say. It could just be an integer. Or
> it could be a type object that's being dispatched on.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/hackerschool/webstack.jl/pull/7/files#r3445018
> .

chuckha replied:
What about [GET] or [POST] for consistency?

zachallaun replied:
I don't believe it's poor style in Julia to have multiple ways to call a function. Defining this form will probably be as simple as

``` jl
function route(handler::Function, app::Micro.App, method::Micro.Method, resource::String)
    route(handler, app, [method], resource)
end
```

StefanKarpinski replied:
If these methods are just integers, you could make them all powers of two and just OR them together instead: `GET | PUT`. Efficiency doesn't much matter for configuration, but it is actually a nice interface. The type could be changed later to something more complicated and still use the same syntax even.

zachallaun replied:
That's… awesome.

despeset committed:
Adds middleware function to Meddle API

middleware function constructs a MidwareStack of Midware types from a list of either Midware or functions that return Midware.

despeset committed:
Changes method syntax to use Stefan's bitwise fun

despeset committed:
Adds half-baked Micro implementation

despeset committed:
Merge branch 'api-speculation'

Conflicts:
	example-app.jl

zachallaun committed:
Moves Micro.jl into directory

zachallaun committed:
Simple Tree and search implementations

This also includes a function for searching Trees composed of routes.

zachallaun committed:
Remove accidentally committed stuffs

despeset committed:
Adds simple routing and working Micro example

Introduces the concept of tree based routing tables.
Routing can only handle "static" routes right now.
Micro now leverages Meddle to build a middleware stack with itself at the end.

zachallaun committed:
Routes and Trees are no longer modules

zachallaun committed:
route function can accept a Function as last arg

zachallaun committed:
Http uses Tasks

Instead of calling handlers in parser callbacks, a new Task is spawned
with `@async`.

zachallaun commented:
This branch has been rebased locally and merged into master.
